stdlib/src/utils/string_slice.mojo

```diff
@@ -978,14 +1155,256 @@ trait Stringlike:
+
+
+@always_inline
+fn to_string_list[
+    O: ImmutableOrigin, //
+](items: List[StringSlice[O]]) -> List[String]:
+    """Create a list of Strings copying the existing data.
+
+    Parameters:
+        O: The origin of the data.
+
+    Args:
+        items: The List of string slices.
+
+    Returns:
+        The list of created strings.
+    """
+
+    fn unsafe_ptr_fn(v: StringSlice[O]) -> UnsafePointer[Byte]:
+        return v.unsafe_ptr()
+
+    fn len_fn(v: StringSlice[O]) -> Int:
+        return v.byte_length()
+
+    return _to_string_list[len_fn, unsafe_ptr_fn](items)
+
+
+@always_inline
+fn to_string_list[
+    O: ImmutableOrigin, //
+](items: List[Span[Byte, O]]) -> List[String]:
+    """Create a list of Strings copying the existing data.
+
+    Parameters:
+        O: The origin of the data.
+
+    Args:
+        items: The List of Bytes.
+
+    Returns:
+        The list of created strings.
+    """
+
+    fn unsafe_ptr_fn(v: Span[Byte, O]) -> UnsafePointer[Byte]:
+        return v.unsafe_ptr()
+
+    fn len_fn(v: Span[Byte, O]) -> Int:
+        return len(v)
+
+    return _to_string_list[len_fn, unsafe_ptr_fn](items)
+
+
+@always_inline
+fn _split[
+    T0: Stringlike, //,
+    has_maxsplit: Bool,
+    has_sep: Bool,
+    T1: Stringlike = StringLiteral,
+](src_str: T0, sep: Optional[T1], maxsplit: Int) -> List[String]:
+    @parameter
+    if has_sep:
+        return to_string_list(
+            _split_impl[has_maxsplit](src_str, sep.value(), maxsplit)
+        )
+    else:
+        return to_string_list(_split_impl[has_maxsplit](src_str, maxsplit))
+
+
+fn _split_sl[
+    O: ImmutableOrigin, //,
+    has_maxsplit: Bool,
+    has_sep: Bool,
+    T: Stringlike = StringLiteral,
+](src_str: StringSlice[O], sep: Optional[T], maxsplit: Int) -> List[
+    StringSlice[O]
+] as output:
+    var items: List[Span[Byte, O]]
+
+    @parameter
+    if has_sep:
+        items = rebind[__type_of(items)](
+            _split_impl[has_maxsplit](src_str, sep.value(), maxsplit)
+        )
+    else:
+        items = rebind[__type_of(items)](
+            _split_impl[has_maxsplit](src_str, maxsplit)
+        )
+
+    i_len = len(items)
+    out_ptr = UnsafePointer[StringSlice[O]].alloc(i_len)
+    i = 0
+    for item in items:
+        (out_ptr + i).init_pointee_move(StringSlice[O](unsafe_from_utf8=item[]))
+        i += 1
+    output = __type_of(output)(
+        unsafe_pointer=out_ptr, size=i_len, capacity=i_len
+    )
+
+
+fn _split_impl[
+    T0: Stringlike,
+    T1: Stringlike,
+    O: ImmutableOrigin, //,
+    has_maxsplit: Bool,
+](ref [O]src_str: T0, sep: T1, maxsplit: Int) -> List[Span[Byte, O]] as output:
+    sep_len = len(sep.as_bytes_read())
+    if sep_len == 0:
+        iterator = src_str.__iter__()
+        i_len = len(iterator) + 2
+        out_ptr = UnsafePointer[Span[Byte, O]].alloc(i_len)
+        out_ptr.init_pointee_move(rebind[Span[Byte, O]]("".as_bytes_read()))
```

Comment:
I think it should be roughly it should take the value `src_str.as_bytes()[:0]` so the `ptr` still points to the original string.

Reply:
great idea, thanks